include/circt/Dialect/FIRRTL/InstanceGraph.h

```diff
@@ -0,0 +1,261 @@
+
+  /// The module.
+  Operation *module;
+
+  /// List of instance operations in this module.  This member owns the
+  /// InstanceRecords, which may be pointed to by other InstanceGraohNode's use
+  /// lists.
+  EdgeVec moduleInstances;
+
+  /// List of instances which instantiate this module.
+  UseVec moduleUses;
+
+  // Provide access to the constructor.
+  friend class InstanceGraph;
+};
+
+/// This graph tracks modules and where they are instantiated. This is intended
+/// to be used as a cached analysis on FIRRTL circuits.  This class can be used
+/// to walk the modules efficiently in a bottom-up or top-down order.
+///
+/// To use this class, retrieve a cached copy from the analysis manager:
+///   auto &instanceGraph = getAnalysis<InstanceGraph>(getOperation());
+class InstanceGraph {
+
+  /// Storage for InstanceGraphNodes.
+  using NodeVec = std::vector<std::unique_ptr<InstanceGraphNode>>;
+
+  /// Iterator that unwraps a unique_ptr to return a regular pointer.
+  static InstanceGraphNode *unwrap(const NodeVec::value_type &value) {
+    return value.get();
+  }
+  struct NodeIterator final
+      : public llvm::mapped_iterator<NodeVec::const_iterator,
+                                     decltype(&unwrap)> {
+    /// Initializes the result type iterator to the specified result iterator.
+    NodeIterator(NodeVec::const_iterator it)
+        : llvm::mapped_iterator<NodeVec::const_iterator, decltype(&unwrap)>(
+              it, &unwrap) {}
+  };
+
+public:
+  /// Create a new module graph of a circuit.  This must be called on a FIRRTL
+  /// CircuitOp.
+  explicit InstanceGraph(Operation *operation);
+
+  /// Get the node corresponding to the top-level module of a circuit.
+  InstanceGraphNode *getTopLevelNode() const;
+
+  /// Look up an InstanceGraphNode for a module. Operation must be an FModuleOp
+  /// or an FExtModuleOp.
+  InstanceGraphNode *lookup(Operation *op) const;
+
+  /// Lookup an InstanceGraphNode for a module. Operation must be an FModuleOp
+  /// or an FExtModuleOp.
+  InstanceGraphNode *operator[](Operation *op) const { return lookup(op); }
+
+  /// Look up the referenced module from an InstanceOp. This will use a
+  /// hashtable lookup to find the module, where
+  /// InstanceOp.getReferencedModule() will be a linear search through the IR.
+  Operation *getReferencedModule(InstanceOp op) const;
+
+  /// Iterate through all modules.
+  using iterator = NodeIterator;
+  iterator begin() const { return nodes.begin(); }
+  iterator end() const { return nodes.end(); }
+
+private:
+  /// Get the node corresponding to the module.  If the node has does not exist
+  /// yet, it will be created.
+  InstanceGraphNode *getOrAddNode(StringRef name);
+
+  /// Lookup an module by name.
+  InstanceGraphNode *lookup(StringRef name) const;
+
+  /// The storage for graph nodes, with deterministic iteration.
+  NodeVec nodes;
+
+  /// This maps each operation to its graph node.
+  llvm::StringMap<unsigned> nodeMap;
+};
```

Comment:
Since you already have a map, you could map string -> InstanceGraphNode.  Do you need insertion-order iteration?  Especially since the NodeVec is of pointers, you don't get any locality from using a vector for the backend storage.  (Same comment as below about using a std::deque to avoid uniq_ptr)

Reply:
Yeah, its implemented to have deterministic iteration order.  I'm not totally sure if I need it, but the other call graph examples in LLVM I was looking at had it.